#  Pandas — Series, DataFrame, eksploracja danych
Autor: Jan Pyka

In [34]:
import pandas as pd
import numpy as np

## Ćwiczenie 1: Series — tworzenie i operacje

### Krok 1

In [33]:
oceny = pd.Series([4.5, 3.0, 5.0, 3.5, 4.0])
print(oceny)

0    4.5
1    3.0
2    5.0
3    3.5
4    4.0
dtype: float64


In [2]:
pensje = pd.Series({'Anna': 5500, 'Jan': 7200, 'Ewa': 4800, 'Marek': 9100, 'Kasia': 6300})
print(pensje)

Anna     5500
Jan      7200
Ewa      4800
Marek    9100
Kasia    6300
dtype: int64


In [ ]:
#1
print(pensje.idxmax())
#2
print(f'Średnia pensja wynosi: {pensje.mean()}')
#3
print(f'Osoby zarabiające powyżej średniej to:\n{pensje[pensje > pensje.mean()]}')
#4
print(f'Liczba osób zarabiająca < 6000 to: {(pensje[pensje < 6000]).count()}')

Marek
Średnia pensja wynosi: 6580.0
Osoby zarabiające powyżej średniej to:
Jan      7200
Marek    9100
dtype: int64
Liczba osób zarabiająca < 6000 to: 2


### Krok 2

In [31]:
#1
pensje_podwyzka = pensje * 1.10
print(f'Podwyzka:\n{pensje_podwyzka}')
#2
pensje_premia = pensje + 500
print(f'Premia:\n{pensje_premia}')
#3
nowa_pensja = (pensje * 1.10) + 500
print(f'Nowa:\n{nowa_pensja}')
#4
print(f'Sort:\n{nowa_pensja.sort_values(ascending=False)}')

Podwyzka:
Anna      6050.0
Jan       7920.0
Ewa       5280.0
Marek    10010.0
Kasia     6930.0
dtype: float64
Premia:
Anna     6000
Jan      7700
Ewa      5300
Marek    9600
Kasia    6800
dtype: int64
Nowa:
Anna      6550.0
Jan       8420.0
Ewa       5780.0
Marek    10510.0
Kasia     7430.0
dtype: float64
Sort:
Marek    10510.0
Jan       8420.0
Kasia     7430.0
Anna      6550.0
Ewa       5780.0
dtype: float64


## Ćwiczenie 2: DataFrame — tworzenie i atrybuty 

### Krok 1

In [35]:
# Dane o produktach sklepu internetowego
sklep = pd.DataFrame({
'produkt': ['Laptop', 'Tablet', 'Smartfon', 'Słuchawki', 'Monitor'],
'cena': [3500, 1800, 2500, 350, 2200],
'stan_magazynowy': [45, 120, 200, 500, 75],
'kategoria': ['Komputery', 'Mobilne', 'Mobilne', 'Akcesoria', 'Komputery']
})


In [53]:
#1
print(f'Ksztalt tabeli: {sklep.shape}')
#2
print(f'Nazwy kolumn: {sklep.columns}')
#3
print(f'Typy danych:\n{sklep.dtypes}')
#4
print(f'Pierwsze 3:\n{sklep.head()}')

Ksztalt tabeli: (5, 4)
Nazwy kolumn: Index(['produkt', 'cena', 'stan_magazynowy', 'kategoria'], dtype='str')
Typy danych:
produkt              str
cena               int64
stan_magazynowy    int64
kategoria            str
dtype: object
Pierwsze 3:
     produkt  cena  stan_magazynowy  kategoria
0     Laptop  3500               45  Komputery
1     Tablet  1800              120    Mobilne
2   Smartfon  2500              200    Mobilne
3  Słuchawki   350              500  Akcesoria
4    Monitor  2200               75  Komputery


In [58]:
sklep['wartosc_magazynu'] = sklep['cena'] * sklep['stan_magazynowy']
sklep['cena_z_vat'] = sklep['cena'] * 1.23
print(sklep)

     produkt  cena  stan_magazynowy  kategoria  wartosc_magazynu  cena_z_vat
0     Laptop  3500               45  Komputery            157500      4305.0
1     Tablet  1800              120    Mobilne            216000      2214.0
2   Smartfon  2500              200    Mobilne            500000      3075.0
3  Słuchawki   350              500  Akcesoria            175000       430.5
4    Monitor  2200               75  Komputery            165000      2706.0


### Krok 2

In [59]:
# Wczytaj dataset tips
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv'
tips = pd.read_csv(url)

In [71]:
#1
print(f'Kolumny i wiersze: {tips.shape}\n')
#2
print(f'Ostatnie 5:\n{tips.tail()}\n')
#3
print(f'Trzy losowe:\n{tips.sample(3)}\n')
#4
print(f'Typy danych:\n{tips.dtypes}')

Kolumny i wiersze: (244, 7)

Ostatnie 5:
     total_bill   tip     sex smoker   day    time  size
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

Trzy losowe:
     total_bill   tip   sex smoker  day    time  size
31        18.35  2.50  Male     No  Sat  Dinner     4
242       17.82  1.75  Male     No  Sat  Dinner     2
207       38.73  3.00  Male    Yes  Sat  Dinner     4

Typy danych:
total_bill    float64
tip           float64
sex               str
smoker            str
day               str
time              str
size            int64
dtype: object


## Ćwiczenie 3: Eksploracja datasetu — samodzielna analiza

In [72]:
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv'
penguins = pd.read_csv(url)

### Zad.3a:

In [80]:
#1
print(f'Head(10):\n{penguins.head(10)}\n')
#2
print(f'Shape: {penguins.shape}')

Head(10):
  species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   
5  Adelie  Torgersen            39.3           20.6              190.0   
6  Adelie  Torgersen            38.9           17.8              181.0   
7  Adelie  Torgersen            39.2           19.6              195.0   
8  Adelie  Torgersen            34.1           18.1              193.0   
9  Adelie  Torgersen            42.0           20.2              190.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  
5       3650.0    MALE  


In [ ]:
#3
print(f'Info: {penguins.info()}')
# W sumie brakuje 19 wartości w komórkach

<class 'pandas.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    str    
 1   island             344 non-null    str    
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    str    
dtypes: float64(4), str(3)
memory usage: 18.9 KB
Info: None


### Zad.3b:

In [ ]:
#1
print(penguins.describe())
#Pingwiny srednio malo roznia sie innymi parametrami poza waga

       bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count      342.000000     342.000000         342.000000   342.000000
mean        43.921930      17.151170         200.915205  4201.754386
std          5.459584       1.974793          14.061714   801.954536
min         32.100000      13.100000         172.000000  2700.000000
25%         39.225000      15.600000         190.000000  3550.000000
50%         44.450000      17.300000         197.000000  4050.000000
75%         48.500000      18.700000         213.000000  4750.000000
max         59.600000      21.500000         231.000000  6300.000000


In [ ]:
#2
print(penguins.describe(include='all'))
#kolumny tekstowe dodaja informacji o gatunkach, wyspach i plci 

#3
#Mediana masy ciala wynosi 4050g

       species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
count      344     344      342.000000     342.000000         342.000000   
unique       3       3             NaN            NaN                NaN   
top     Adelie  Biscoe             NaN            NaN                NaN   
freq       152     168             NaN            NaN                NaN   
mean       NaN     NaN       43.921930      17.151170         200.915205   
std        NaN     NaN        5.459584       1.974793          14.061714   
min        NaN     NaN       32.100000      13.100000         172.000000   
25%        NaN     NaN       39.225000      15.600000         190.000000   
50%        NaN     NaN       44.450000      17.300000         197.000000   
75%        NaN     NaN       48.500000      18.700000         213.000000   
max        NaN     NaN       59.600000      21.500000         231.000000   

        body_mass_g   sex  
count    342.000000   333  
unique          NaN     2  
top

### Zad.3c:

In [94]:
#1
print(f'Pingwiny kazdego gatunku:\n{penguins.value_counts(penguins.species)}\n')
#2
print(f'Pingwiny na kazdej wyspie:\n{penguins.value_counts(penguins.island)}\n')
#3
print(f'Rozklad plci:\n{penguins.value_counts(penguins.sex)}\n')
#4
print(f'Rozklad gatunkow:\n{penguins.value_counts(penguins.species,normalize=True)}\n')

Pingwiny kazdego gatunku:
species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

Pingwiny na kazdej wyspie:
island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64

Rozklad plci:
sex
MALE      168
FEMALE    165
Name: count, dtype: int64

Rozklad gatunkow:
species
Adelie       0.441860
Gentoo       0.360465
Chinstrap    0.197674
Name: proportion, dtype: float64



### Zad.3d:

In [108]:
#1
print(penguins.isna().sum())
#2
print((penguins.sex.isna().sum()) / len(penguins * 100))

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64
0.03197674418604651


### Zad.3e:

Dataset zawiera 344 pingwinów z 3 gatunków <br>
Najczęstszy gatunek to Adelie (152 osobników, 44%)<br>
Brakuje danych w kolumnach bill_length_mm, bill_depth_mm, flipper_length_mm, body_mass_g, sex (łącznie 19 braków)<br>
Średnia masa ciała to 4201 g (mediana: 4050 g)<br>
Pingwiny mieszkają na 3 wyspach, najliczniej na Biscoe

## Ćwiczenie 4: Pytania biznesowe

### Zad.4a:

In [109]:
tips = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv')
# Dodaj kolumnę z procentem napiwku
tips['tip_pct'] = (tips['tip'] / tips['total_bill'] * 100).round(1)

In [127]:
# 1. Oblicz średni rachunek (total_bill)
print(f'Sredni rachunek: {tips.total_bill.mean()}')
# 2. Oblicz średni napiwek w procentach (tip_pct)
print(f'Sredni napiwek w proc: {tips.tip_pct.mean()}\n')
# 3. Wyświetl który dzień ma najwięcej rachunków (value_counts na 'day') (Sat)
print(f'{tips.day.value_counts()}\n')
# 4. Policz ile rachunków jest od palaczy (value_counts na 'smoker') (93)
print(f'{tips.smoker.value_counts()}\n')
# 5. Wyświetl describe() tylko dla kolumny 'tip' — jaka jest mediana napiwku? (2.9)
print(f'{tips.describe()}')

Sredni rachunek: 19.78594262295082
Sredni napiwek w proc: 16.08155737704918

day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64

smoker
No     151
Yes     93
Name: count, dtype: int64

       total_bill         tip        size     tip_pct
count  244.000000  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672   16.081557
std      8.902412    1.383638    0.951100    6.107449
min      3.070000    1.000000    1.000000    3.600000
25%     13.347500    2.000000    2.000000   12.875000
50%     17.795000    2.900000    2.000000   15.450000
75%     24.127500    3.562500    3.000000   19.125000
max     50.810000   10.000000    6.000000   71.000000


In [138]:
# 6. Jaki procent rachunków pochodzi z obiadu (Dinner)? (72%)
print(f'{tips.time.value_counts(normalize=True)}\n')
# Podpowiedź: value_counts(normalize=True) na kolumnie 'time'
# 7. Ile unikalnych rozmiarów grup jest w danych? (nunique na 'size')
print(f'Unikalnych rozmiarow grup: {tips['size'].nunique()}')

time
Dinner    0.721311
Lunch     0.278689
Name: proportion, dtype: float64

Unikalnych rozmiarow grup: 6
